palette/src/js/script.js

```diff
@@ -0,0 +1,409 @@
+    const fillColor = Palette.colorToRgba(this.currentColor);
+    const dstImg = this.ctx.getImageData(0, 0, this.scale, this.scale);
+    const dstData = dstImg.data;
+    const startPos = this.getPixelPos(startX, startY);
+    let startColor = {
+      r: dstData[startPos],
+      g: dstData[startPos + 1],
+      b: dstData[startPos + 2],
+      a: dstData[startPos + 3],
+    };
+    if (fillColor.r === startColor.r
+      && fillColor.g === startColor.g
+      && fillColor.b === startColor.b
+    ) {
+      startColor = {
+        r: 0,
+        g: 0,
+        b: 0,
+        a: 0,
+      };
+    }
+    const todo = [[startX, startY]];
+    while (todo.length) {
+      const pos = todo.pop();
+      const x = pos[0];
+      let y = pos[1];
+      let currentPos = this.getPixelPos(x, y);
+      y += 1;
+      while ((y >= 0) && Palette.matchStartColor(dstData, currentPos, startColor)) {
+        y -= 1;
+        currentPos -= this.scale * 4;
+      }
+      currentPos += this.scale * 4;
+      let reachLeft = false;
+      let reachRight = false;
+      y -= 1;
+      while ((y < this.scale - 1) && Palette.matchStartColor(dstData, currentPos, startColor)) {
+        y += 1;
+        Palette.colorPixel(dstData, currentPos, fillColor);
+        if (x > 0) {
+          if (Palette.matchStartColor(dstData, currentPos - 4, startColor)) {
+            if (!reachLeft) {
+              todo.push([x - 1, y]);
+              reachLeft = true;
+            }
+          } else if (reachLeft) {
+            reachLeft = false;
+          }
+        }
+        if (x < this.scale - 1) {
+          if (Palette.matchStartColor(dstData, currentPos + 4, startColor)) {
+            if (!reachRight) {
+              todo.push([x + 1, y]);
+              reachRight = true;
+            }
+          } else if (reachRight) {
+            reachRight = false;
+          }
+        }
+        currentPos += this.scale * 4;
+      }
+    }
+    this.ctx.putImageData(dstImg, 0, 0);
+    return true;
+  }
+
+  // convert format color
+  static rgbToHex(data) {
+    let r = data.r.toString(16);
+    let g = data.g.toString(16);
+    let b = data.b.toString(16);
+    const { a } = data;
+    if (a === 0) {
+      r = 'ff';
+      g = 'ff';
+      b = 'ff';
+    } else {
+      if (r.length === 1) r = `0${r}`;
+      if (g.length === 1) g = `0${g}`;
+      if (b.length === 1) b = `0${b}`;
+    }
+    return `#${r}${g}${b}`;
+  }
+
+  colorPicker(data) {
+    const startX = this.getCoords(data).left;
+    const startY = this.getCoords(data).top;
+    const dstImg = this.ctx.getImageData(0, 0, this.scale, this.scale);
+    const dstData = dstImg.data;
+    const startPos = this.getPixelPos(startX, startY);
+    const startColor = {
+      r: dstData[startPos],
+      g: dstData[startPos + 1],
+      b: dstData[startPos + 2],
+      a: dstData[startPos + 3],
+    };
+    const newColor = Palette.rgbToHex(startColor);
+    console.log(startColor);
+    console.log(newColor);
```

Comment:
Don't forget to remove all the `console.log` and commented code once you commit the final version